lib/virt_autotest/kernel.pm

```diff
@@ -42,6 +42,12 @@ sub check_virt_kernel {
     record_info "KERNEL $target$suffix", "We are now checking kernel on $target$suffix.";
     assert_script_run qq(echo -e "\\n# $target$suffix:" >> $log_file);
 
+    assert_script_run("$go_to_target grep -i ^welcome /etc/issue | tee -a $log_file");
+    if ($sles_running_version >= 12) {
+        script_run("$go_to_target journalctl -b | grep -i 'Hypervisor Detected' | tee -a $log_file");
+    } else {
+        script_run("$go_to_target dmesg | grep -i 'Hypervisor Detected' | tee -a $log_file");
+    }
```

Comment:
Our tests should always consider openSUSE Tumbleweed as the default product without a version. That means any differing behaviour should have an explicit exclude rule for the "older products" and potentially an explanation why the behaviour should differ. You can try to negate the logic check.

See https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/CONTRIBUTING.md#coding-style for more details

Reply:
@okurz hello Oliver, do we have bare metal workers on O3? I mean I of course agree. But those virtualization tests are currently very SLE specific.

Reply:
That was not the point. It's about how to write the code. Very simple instead of

```perl
if (sle >= 12) {
   # new stuff
}
else {
   # old stuff
}
```

write

```perl
if (sle < 12) {
   # old stuff
}
else {
   # new stuff
}
```

this way you are not missing out any potential new version (like Tumbleweed or Leap) that also needs the new behaviour and it's more obvious what is backward-compatibility code that can be deleted as soon as "sle < 12" goes EOL.

Reply:
I got your point. But is mine current solution wrong?

Reply:
hm, AFAICS you are not using the distinction anymore and use a more generic solution which of course is even better. I assume that I have not overlooked something so we can mark the conversation as resolved.